File: 145-BinaryTreePostorderTraversal/145-BinaryTreePostorderTraversal.py
# Last updated: 2/28/2026, 4:28:21 PM
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        
        if not root:
            return []

        stack, output = [root], []

        while stack:
            node = stack.pop()
            output.append(node.val)
   
            if node.left:
                stack.append(node.left)
            if node.right:
                stack.append(node.right)

      
        return output[::-1]
        
        

File: 145-BinaryTreePostorderTraversal/145-BinaryTreePostorderTraversal.java
// Last updated: 2/28/2026, 4:28:22 PM
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {

        if(root == null) return new ArrayList<Integer>();

        Stack<TreeNode> stack1 = new Stack<>();
        Stack<TreeNode> stack2 = new Stack<>();

        stack1.push(root);

        while(!stack1.isEmpty()){

            TreeNode node = stack1.pop();
            stack2.push(node);

            if(node.left != null){
                stack1.push(node.left);
            }

            if(node.right != null){
                stack1.push(node.right);
            }

        }

        List<Integer> postorder = new ArrayList<>();

        while(!stack2.isEmpty()){
            postorder.add(stack2.pop().val);
        }

        return postorder;
        
    }
}